_LowLevelCode/cpp/cpp_communicator/MPI_wrapper.h

```diff
@@ -70,13 +78,21 @@ class MPI_wrapper {
         return &this->asyncMessList;
     }
     // get access to the synchroneous messages holder.
-    SendMessHolder* get_sync_queue() {
-        return &this->SyncMessHolder;
+    SendMessHolder* get_sync_queue(int dest_address = 0) {
+        return &this->SyncMessHolder[dest_address];
+    }
+    // get access to the interrupt holder
+    SendMessHolder *get_interrupt_queue(int dest_address = 0) {
+        return &this->InterruptHolder[dest_address];
     }
     // check if any message present in test mode
     bool any_message_present() {
-        if (SyncMessHolder.theRequest==0) 
-            return true;
+        for (auto it = InterruptHolder.begin(); it != InterruptHolder.end(); it++) {
+            if (it->theRequest == 0)  return true;
+        }
```

Comment:
The standard way of doing this would be with a range-based for loop:

```suggestion
        for (auto &msg_holder : InterruptHolder) {
            if (msg_holder.theRequest == 0)  
                return true;
        }
```

Reply:
may be. Hate references from FORTRAN times. Does this mean that every variable in iteration is copied into `msg_holder` or not? Its not important in this case, as this code is for tests only, but hate using the code, not fully understanding what is happening

Reply:
No, in this case nothing would be copied. `msg_holder` will simply point to the items in `InterruptHolder`.

The `&` means that nothing is copied.

```cpp
for (auto msg_holder : InterruptHolder) { ... }  // msg_holder copies items in InterruptHolder
for (auto &msg_holder : InterruptHolder) { ... }  // msg_holder points to items in InterruptHolder
```

Reply:
In fact, since we're not editing `msg_holder` the loop should really be `for (const auto &msg_holder) { ... }`. I've changed the suggestion above to reflect this.

Reply:
Can we please have brackets around the `return true;` in the `if` too

Reply:
not sure its entirely true because 
`
   for (auto &&it : InterruptHolder) {
            if (it.theRequest == 0)  return true;
        }
`
works absolutely fine too. And its not a pointer.... Will try const 
@nickbattam-tessella  -- I've changed it to one-liner -- looks better without the brackets.

Reply:
I wasn't completely accurate with what I said before `for (auto &msg_holder : InterruptHolder)` doesn't make `msg_holder` a `SendMessHolder`, it's an iterator (a `std::vector<T>::iterator`, which is what `InterruptHolder.begin()` returns). 

The difference between `for (auto &x : vec ) {}` and `for (auto &&x : vec) {}` is explained here https://stackoverflow.com/questions/29320575/whats-the-difference-between-and-in-a-range-based-for-loop

Reply:
though, according to the article, most performant way of doing requests would be && as its direct address of the data. 
const & may default to iterator.

Reply:
`auto &&` does not give `const` correctness. If you only want read-access to the items in the container, you should use `const`